source/slang/slang-ir-autodiff-unzip.h

```diff
@@ -66,6 +86,185 @@ struct DiffUnzipPass
+            {
+                auto primalArg = lookupPrimalInst(arg);
+                auto diffArg = lookupDiffInst(arg);
+
+                // If arg is a mixed differential (pair), it should have already been split.
+                SLANG_ASSERT(primalArg);
+                SLANG_ASSERT(diffArg);
+
+                auto pairArg = diffBuilder->emitMakeDifferentialPair(
+                        arg->getDataType(),
+                        primalArg,
+                        diffArg);
+
+                diffBuilder->markInstAsDifferential(pairArg, primalArg->getDataType());
+                diffArgs.add(pairArg);
+            }
+            else
+            {
+                diffArgs.add(arg);
+            }
+        }
+        
+        auto newFwdCallee = diffBuilder->emitForwardDifferentiateInst(fwdCalleeType, baseFn);
+        diffBuilder->markInstAsDifferential(newFwdCallee);
+
+        auto diffPairVal = diffBuilder->emitCallInst(
+            fwdPairResultType, 
+            newFwdCallee,
+            diffArgs);
+        diffBuilder->markInstAsDifferential(diffPairVal, primalType);
+
+        auto diffVal = diffBuilder->emitDifferentialPairGetDifferential(diffType, diffPairVal);
+        diffBuilder->markInstAsDifferential(diffVal, primalType);
+
+        return InstPair(primalVal, diffVal);
+    }
+
+    InstPair splitMakePair(IRBuilder*, IRBuilder*, IRMakeDifferentialPair* mixedPair)
+    {
+        return InstPair(mixedPair->getPrimalValue(), mixedPair->getDifferentialValue());
+    }
+
+    InstPair splitLoad(IRBuilder* primalBuilder, IRBuilder* diffBuilder, IRLoad* mixedLoad)
+    {
+        // By the nature of how diff pairs are used, and the fact that FieldAddress/GetElementPtr,
+        // etc, cannot appear before a GetDifferential/GetPrimal, a mixed load can only be from a
+        // parameter or a variable.
+        // 
+        if (as<IRParam>(mixedLoad->getPtr()))
+        {
+            // Should not occur with current impl of fwd-mode.
+            // If impl. changes, impl this case too.
+            // 
+            SLANG_UNIMPLEMENTED_X("Splitting a load from a param is not currently implemented.");
+        }
+
+        // Everything else should have already been split.
+        auto primalPtr = lookupPrimalInst(mixedLoad->getPtr());
+        auto diffPtr = lookupDiffInst(mixedLoad->getPtr());
+
+        return InstPair(primalBuilder->emitLoad(primalPtr), diffBuilder->emitLoad(diffPtr));
+    }
+
+    InstPair splitVar(IRBuilder* primalBuilder, IRBuilder* diffBuilder, IRVar* mixedVar)
+    {
+        auto pairType = as<IRDifferentialPairType>(mixedVar->getDataType());
+        auto primalType = pairType->getValueType();
+        auto diffType = (IRType*) diffTypeContext.getDifferentialForType(primalBuilder, primalType);
+        
+        return InstPair(primalBuilder->emitVar(primalType), diffBuilder->emitVar(diffType));
+    }
+
+    InstPair splitReturn(IRBuilder*, IRBuilder* diffBuilder, IRReturn* mixedReturn)
```

Comment:
Could probably use some comments to explain why splitting a return will result in returning a pair that is marked as differential here.

Reply:
Noted! I'll add more comments to all `splitXXX()` to make the intent clearer.